Update keyfiguresfinland

tuopouk opened the pull request:
Hi @AnnMarieW 

I went through the code and tried to hit most of the checkmarks.  I basically minimized everything by removing default parameters and replaced style parameters with classnames. The wrangling part in the beginning might be difficult to understand but it works.

I also changed the json file since for some reason Åland was not visible on the map. Now it works.

I replaced orjson with json since orjson is not in the requirements.txt. Orjson might faster though. 

I've been running through issues occasionally with mapboxchoropleth. Sometimes the maps do not render due to javascript errors. It might be issue with browser plugins or for some reason the plotly component cannot retrieve the map data from mapbox. Also I've noticed that having a mapbox token makes this error less-occurring. I wonder if we have a chance to add a token in the code. This app works on my virtual environment with the requirements.txt from the root.

AnnMarieW commented:
Thanks for doing the update!

I think the data wrangling at the start is fine.  People would replace that if they use their own data source, so it still checks the "app is extendable"  box.  

I'm not opposed to adding `orjson`, if it's simple.  From my understanding, just having `orjson` in the environment speeds things up.  Would you like to see if that's the case?

It could be really helpful to have an example app with a mapbox token. However, we should make sure we can stay in the free tier in production.

Before we continue with the review, it's important that the app runs error free.  Just to be clear, are you still seeing errors occasionally even with a token?

tuopouk commented:
@AnnMarieW 
Yeah this app works just fine on production. I deployed it to [Heroku](https://maptestheroku.herokuapp.com/) and it works just fine. 

I've been expanding this project a bit and I have a version where you can select map style and use hover data to visualize time series of a key figure. I have a version on [Heroku ](https://keyfiguresfinland.herokuapp.com/) with a general token and a version on [Render](https://keyfiguresfinland.onrender.com/) without any token. With the Render version I tried adding the token and it worked. I'm just saying adding the token gives a bit certainty on rendering since it is not a Dash/Python issue. I guess the token should not be defined in the app code. But anyways, this app works well even without the token on a virtual environment and on Heroku. 

I'm also fine either ways adding or not adding orjson. In this case it does not make a big difference since the json file is loaded once in the beginning and is not huge. It might speed up but I doubt you could see the difference here.

AnnMarieW commented:
Ok - sounds good!  Your site looks AWESOME :star_struck:   We will definitely put a link to it from the Example Index.  It will be great to show how a minimal example can grow into a full featured site.    And your site is a fantastic way to try out other themes, map types, figure templates, etc.  

In the meantime, I'd like to tweak the styling of the version that will go in Example Index.  Here's a version to make it work on various screen sizes.  I'll make some inline comments in the code, and you can give it a try.


![image](https://user-images.githubusercontent.com/72614349/201481241-b028fb7c-f428-4d2e-9bbc-b7e24a199e0a.png)

AnnMarieW commented on examples/keyfiguresfinland.py:
```diff
@@ -0,0 +1,136 @@
+                dcc.Dropdown(id = 'key-figures-finland-key-figure-selection-x',
+                             options = sorted(list(pd.unique(df.columns))),
+                             value = "Degree of urbanisation, %, 2020",
+                             className = 'text-dark bg-light text-nowrap'
```
Can remove "text-dark bg-light" because you can't override the colors this way for the dcc.Dropdown

tuopouk replied:
That looks better on md. I'm not that much into styling so I don't how you think about the positioning of the map in md screens now. The color scale bar seems to move it a bit too much on the left and thus part of the country goes off the screen. Maybe the zoom level map center should be redefined. I tried with  `'zoom':3.8,'center':{'lat': 64.961093, 'lon': 25.795386}` and I find it a bit better.

AnnMarieW commented on examples/keyfiguresfinland.py:
```diff
@@ -0,0 +1,136 @@
+                             value = "Degree of urbanisation, %, 2020",
+                             className = 'text-dark bg-light text-nowrap'
+                             ),                
+                html.H1(id = 'key-figures-finland-whole-country-header-x', className="mt-5 display-2 text-center"),
```
Should change this to an `html.H2` so it's consistent with the "display-2"

AnnMarieW commented on examples/keyfiguresfinland.py:
```diff
@@ -0,0 +1,136 @@
+    function(geojson, locations, z){           
+       
+        var layout = {
+            'height':800,
```
Change to `'height':600` so it matches the size of the data in the left column

AnnMarieW commented on examples/keyfiguresfinland.py:
```diff
@@ -0,0 +1,136 @@
+                html.H1(id = 'key-figures-finland-whole-country-header-x', className="mt-5 display-2 text-center"),
+                html.Div(['Data by ',html.A('Statistics Finland', href = 'https://pxdata.stat.fi/PxWeb/pxweb/en/Kuntien_avainluvut/Kuntien_avainluvut__2021/kuntien_avainluvut_2021_viimeisin.px/', target = '_blank')], className="text-center fs-3 text"),
+                
+                ], xs = 12, sm = 12, md = 12, lg = 6, xl = 6, xxl = 6, align = 'center'),
```
It would look better in the example index to have md=6.  Note this can be simplified:
from 
```
xs = 12, sm = 12, md = 12, lg = 6, xl = 6, xxl = 6
```

to:
```
md=6
```
This is because it applies to all screen sizes at the breakpoint and larger.  And the default is 12 so you don't have to set the smaller ones

AnnMarieW commented on examples/keyfiguresfinland.py:
```diff
@@ -0,0 +1,136 @@
+                html.H1(id = 'key-figures-finland-header-x', className="mb-3 mt-3 display-3 text-center"),
+                dcc.Graph(id = 'key-figures-finland-region-map-x', figure = px.choropleth_mapbox(center = {"lat": 64.961093, "lon": 27.590605}))
+                
+                ], xs = 12, sm = 12, md = 12, lg = 6, xl = 6, xxl = 6)
```
Can be changed to `md=6`

AnnMarieW commented on examples/keyfiguresfinland.py:
```diff
@@ -0,0 +1,136 @@
+        dcc.Store(id = 'key-figures-finland-locations-x'),
+        dcc.Store(id = 'key-figures-finland-zs-x'),
+        dcc.Store(id = 'key-figures-finland-geojson-data', data = regions_json),
+        ], fluid = True, className = "dbc")    
```
Can remove `className="dbc"` since that will only work if you are using the styesheet from the dash-bootstrap-templates

AnnMarieW commented on examples/keyfiguresfinland.py:
```diff
@@ -0,0 +1,136 @@
+                
+                ], xs = 12, sm = 12, md = 12, lg = 6, xl = 6, xxl = 6)
+
+            ], justify = 'center', className = "m-auto d-flex justify-content-center"),
```
Can remove 
```
className = "m-auto d-flex justify-content-center"),
```
since it's the same as `justify="center"

AnnMarieW commented on examples/keyfiguresfinland.py:
```diff
@@ -0,0 +1,136 @@
+                ], xs = 12, sm = 12, md = 12, lg = 6, xl = 6, xxl = 6, align = 'center'),
+                
+            dbc.Col([
+                html.H1(id = 'key-figures-finland-header-x', className="mb-3 mt-3 display-3 text-center"),
```
This title over the figure can be removed.  Unlike your full featured site on Heroku, this example doesn't allow you to change from regional to municipal etc.  So the title does not add any new information.  
Note that the callback that updates this title can also be removed

tuopouk replied:
Yeah, makes sense :-)

tuopouk commented:
@AnnMarieW Thank you for the comments. I did not know I was using unnecessary `className`s  that much :-)  
I made styling corrections and removed the header + callback. I also changed the zoom and center of the map.

AnnMarieW commented on examples/keyfiguresfinland.py:
```diff
@@ -0,0 +1,131 @@
+
+# Data provided by Statistics Finland.
+# Original data source: https://pxdata.stat.fi/PxWeb/pxweb/en/Kuntien_avainluvut/Kuntien_avainluvut__2021/kuntien_avainluvut_2021_viimeisin.px/
+df = pd.read_csv('assets/key_figures_regions.csv', encoding = 'latin-1').rename(columns ={'Region 2021':'Region'}).set_index('Region')
```
Why `latin-1` encoding?

tuopouk replied:
Well I've noticed that these Statistics Finland's files require that. Might be an old habit though since I tend to work with files that have nordic alphabet :-)

tuopouk replied:
This actually does also have nordic alphabet since there are ä, ö, and å in region names

AnnMarieW commented on examples/keyfiguresfinland.py:
```diff
@@ -0,0 +1,131 @@
+@app.callback(Output('key-figures-finland-whole-country-header-x','children'),Input('key-figures-finland-key-figure-selection-x', 'value'))
+def update_whole_country_header(key_figure):
+    
+    # Get all the header components for the header.
+    dff = whole_country_df.loc[key_figure]
+    stat_name = dff.stat_name
+    stat_unit = dff.unit
+    stat_year = dff.year
+    stat_value = dff.value
+ 
+    # Change values with no decimals (.0) to int.
+    stat_value = {True: int(stat_value), False: stat_value}['.0' in str(stat_value)]
+    # Use space as thousand separator.
+    stat_value = "{:,}".format(stat_value).replace(',',' ')
+    
+    return html.Div(
+        [
+            html.Div([stat_name, ", ", stat_year]),
+            html.Div("in Finland"),
+            html.Span([stat_value, " ", stat_unit]),
+        ]
+    )
```
This could be made a lot more concise:

```suggestion
 dff = whole_country_df.loc[key_figure]

    # Change values with no decimals (.0) to int.
    stat_value = {True: int(dff.value), False: dff.value}['.0' in str(dff.value)]
    # Use space as thousand separator.
    stat_value = "{:,}".format(stat_value).replace(',', ' ')

    return html.Div(
        [
            html.Div([dff.stat_name, ", ", dff.year]),
            html.Div("in Finland"),
            html.Span([stat_value, " ", dff.unit]),
        ],
    )
```

tuopouk replied:
Thanks for the comments, I'll added the fixes on the newly renamed file

AnnMarieW reviewed (commented):
The map looks better with your adjustment. 

Please change the file name to "key-figures-finland.py", then change the id's so it follow the pattern `<filename>-x-id`,   , so for example, 
`'key-figures-finland-region-map-x'` 
would be:
`'key-figures-finland-x-region-map'`


You can also now delete `keyfiguresfinland_serverside.py`  (thanks for making the version to show the performance difference) 

Additional comments in-line

tuopouk commented on examples/keyfiguresfinland.py:
```diff
@@ -0,0 +1,131 @@
+            dbc.Col([
+                dcc.Graph(id = 'key-figures-finland-region-map-x', figure = px.choropleth_mapbox(center = {"lat": 64.961093, "lon": 25.795386}))
+                
+                ], xs = 12, sm = 12, md = 6, lg = 6, xl = 6, xxl = 6)
```
I see, I did not know that it is enough to just define `md=6`

AnnMarieW replied:
Yes - it's a nice shortcut.  The width applies to the breakpoint and higher - so in this case any screen sizes >= md will have 6 columns.  The default is 12, so there is no need to set the smaller ones.
